katello-attach-subscription

```diff
@@ -99,24 +135,769 @@ if @yaml.has_key?(:settings) and @yaml[:settings].is_a?(Hash)
+    virtwho_location = `which virt-who`
+
+    virtwho_location = virtwho_location.chop!
+    # if which command exit and exit with code 0 then virt-who is installed in virtwho_location
+    if $?.exited? and $?.exitstatus.to_i == 0
+      if @options[:debug]
+        puts "  DEBUG: virt-who installed in #{virtwho_location}"
+      end
+      # now we have to check the version of virt-who launching
+      # dummy versions as virt-who --version is available only from newer versions
+      virtwho_version_number = 0.16
+      # if virt-who it's of the correct version set true the run_virtwho_print flag
+      if virtwho_version_number >= 0.16
+        run_virtwho_print = true
+        # if @options[:usecache] is enabled, before launch virt-who --print we try to read the cache file
+        if @options[:usecache]
+          run_virtwho_print = false
+          # if the virt-who cache file exists, let's open it
+          if File.file?(virtwho_cachefile)
+            begin
+              # first, we read the raw data from the file and put them into raw_json
+              raw_json = File.read(virtwho_cachefile)
+              # then we parse the JSON data to a ruby hash map called json_hash
+              json_hash = JSON.parse(raw_json)
+            rescue Exception
+              abort "FATAL ERROR: Failed to read virt-who cache file for Hypervisors data. Re-run without --read-from-cache."
+            end
+          else
+            puts "FATAL ERROR: --read-from-cache and --virt-who option requested, but the cache file is not currently present."
+            exit 5
+          end
+
+          if @options[:debug]
+            puts "DEBUG: raw virt-who.cache parsed data"
+            p json_hash
+          end
+          # if json_hash is not an hash or the hashmap hasn't any key named hypervisors or has 0 entry in the hashmap then run the virt-who --print
+          if json_hash.is_a?(Hash) and json_hash.has_key?("hypervisors") and json_hash["hypervisors"].count > 0
+            # check the sanity and if every hypervisor have the data we need
+            puts "Starting virt-who cache file parsing. Please be patient."
+            @parsed_hypervisors_array = []
+            json_hash["hypervisors"].each do |json_hypervisor|
+              if @options[:debug]
+                puts " DEBUG: Current Hypervisor fetched data:"
+                p json_hypervisor
+              end
+              # if an entry has valid data push to parsed hypervisor array
+              # check if hypervisor has key facts that contain a valid positive integer value relative the number of socket(s)
+              if json_hypervisor.is_a?(Hash) and json_hypervisor.has_key?("facts") and json_hypervisor["facts"].is_a?(Hash) and ((json_hypervisor["facts"].has_key?("cpu.cpu_socket(s)") and json_hypervisor["facts"]["cpu.cpu_socket(s)"].to_i > 0) or (json_hypervisor["facts"].has_key?("hypervisor.cluster") and json_hypervisor["facts"]["hypervisor.cluster"]))
```

Comment:
in ruby, accessing a non-existent value of a hash yields `nil`, and `nil.to_i` is `0`, so this line can be made shorter w/o introducing bugs :)